[12.0][IMP] account_asset_management: Add analytic tags and propagate

joao-p-marques opened the pull request:
Add analytic tags to `asset.profile` and `asset.asset` and propagate them between these models and to the created `account.move.line`
Add tests for propagation of analytic data from asset to account.move.line

@Tecnativa
TT28974

ping @pedrobaeza @victoralmau

pedrobaeza commented on account_asset_management/tests/account_asset_test_data.xml:
```diff
@@ -78,6 +78,8 @@
       <field name="salvage_value" eval="2000.0"/>
       <field name="profile_id" ref="account_asset_profile_car_5Y"/>
+      <field name="account_analytic_id" ref="analytic.analytic_administratif" />
+      <field name="analytic_tag_ids" search="[('id', 'in', [ref('analytic.tag_contract')])]" />
```
```suggestion
      <field name="analytic_tag_ids" eval="[(4, ref('analytic.tag_contract')]" />
```

pedrobaeza replied:
Sorry, missing ending `)`

pedrobaeza commented on account_asset_management/tests/test_account_asset_management.py:
```diff
@@ -186,7 +186,20 @@ def test_01_nonprorata_basic(self):
+        created_move_ids = vehicle0.depreciation_line_ids[1].create_move()
+        for move_id in created_move_ids:
+            move = self.env["account.move"].browse(move_id)
+            self.assertEquals(
+                len(
+                    move.line_ids.filtered(
+                        lambda line: line.analytic_account_id ==
+                        self.env.ref("analytic.analytic_administratif") and
+                        line.analytic_tag_ids ==
+                        self.env.ref("analytic.tag_contract")
+                    )
+                ),
+                1,
+            )
```
```suggestion
            self.assertTrue(
                move.line_ids.filtered(
                    lambda line: line.analytic_account_id ==
                    self.env.ref("analytic.analytic_administratif") and
                    line.analytic_tag_ids ==
                    self.env.ref("analytic.tag_contract")
                 )
            )
```

joao-p-marques replied:
Thanks for the suggestion @pedrobaeza 
However, this check with the evaluation of the `len()` also asserted that the analytic data was not being propagated to the lines that it shouldn't, but only to the ones of type "expense": https://github.com/OCA/account-financial-tools/pull/1149/files#diff-5318cd3560cdb2d23d2889fe39c5b71745c140810b20f64f0c9aa5b24b8b602bR200-R204
We will lose that with this change... Is that a problem?

pedrobaeza replied:
You are not testing that, but testing that there's only line with such account. If you want to test that, filter by account type, and then `assertEqual` its content.

joao-p-marques replied:
Ok, thanks! I made the change. I now test that the lines with analytic data exist (`assertTrue`) and that they match the appropriate lines, as you suggested (`assertEqual`)

joao-p-marques committed:
[IMP] account_asset_management: Add analytic tags and propagate

Add tests for propagation of analytic data from asset to account.move.line

TT28974

pedrobaeza commented on account_asset_management/tests/test_account_asset_management.py:
```diff
@@ -186,7 +186,24 @@ def test_01_nonprorata_basic(self):
+        created_move_ids = vehicle0.depreciation_line_ids[1].create_move()
+        for move_id in created_move_ids:
+            move = self.env["account.move"].browse(move_id)
+            move_lines_w_analytic_data = move.line_ids.filtered(
+                lambda line: line.analytic_account_id ==
+                self.env.ref("analytic.analytic_administratif") and
+                line.analytic_tag_ids ==
+                self.env.ref("analytic.tag_contract")
+            )
+            self.assertTrue(move_lines_w_analytic_data)
+            self.assertEquals(
+                move_lines_w_analytic_data,
+                move.line_ids.filtered(
+                    lambda line: line.account_id == self.env.ref("account.a_expense")
+                    # Filter for expense lines
+                    # (associated with the profile's depreciation account)
+                )
+            )
```
Cleaner code IMO is:

```suggestion
            expense_line = move.line_ids.filtered(
                lambda line: line.account_id == self.env.ref("account.a_expense")
            )
            self.assertEqual(
                expense_line.analytic_account_id,
                self.env.ref("analytic.analytic_administratif"),
            )
            self.assertEqual(
                expense_line.analytic_tag_ids,
                self.env.ref("analytic.tag_contract")
            )
```

joao-p-marques replied:
Thanks :+1: 
Much better, indeed :smile:

pedrobaeza commented:
/ocabot merge minor

pedrobaeza commented:
Now to v13
